Step 1: object support for nrtsearch(non nested)

taoyyu opened the pull request:
This is the first step to fully support nested objects.
I also modified the int & long parser slightly since 1.0 will fail for int. Also Gson would treat all number as float/double

I can add more tests later

aprudhomme commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/LongFieldDef.java:
```diff
@@ -118,4 +118,8 @@ public Query getTermQuery(TermQuery termQuery) {
   }
+
+  protected Number parseNumberString(String numberString) {
+    return DOUBLE_PARSER.apply(numberString).longValue();
```
Could this lead to precision loss for very large long values? Maybe do this if contains '.'?

aprudhomme commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/IntFieldDef.java:
```diff
@@ -114,4 +114,13 @@ public Query getTermQuery(TermQuery termQuery) {
+
+  @Override
+  protected Number parseNumberString(String numberString) {
+    if (numberString.indexOf('.') == -1) {
+      return super.parseNumberString(numberString);
+    } else {
+      return DOUBLE_PARSER.apply(numberString).intValue();
```
nit: it would be good to have a comment explaining this.

aprudhomme commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/LongFieldDef.java:
```diff
@@ -118,4 +118,12 @@ public Query getTermQuery(TermQuery termQuery) {
   }
+
+  protected Number parseNumberString(String numberString) {
+    if (numberString.indexOf('.') == -1) {
+      return super.parseNumberString(numberString);
+    } else {
+      return DOUBLE_PARSER.apply(numberString).longValue();
```
Same as above.

aprudhomme commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/ObjectFieldDef.java:
```diff
@@ -0,0 +1,92 @@
+    for (Map.Entry<String, IndexableFieldDef> childField : this.getChildFields().entrySet()) {
+      String[] keys = childField.getKey().split("\\.");
+      String key = keys[keys.length - 1];
+      if (childField.getValue().getType().equals("object")) {
```
I think this will need to be capitalized too.

aprudhomme commented on src/test/resources/field/registerFieldsObject.json:
```diff
@@ -0,0 +1,30 @@
+      "search": true,
+      "storeDocValues": true
+    },
+    {
+      "name": "delivery_areas",
+      "type": "OBJECT",
+      "search": true,
+      "childFields": [
+        {
+          "name": "hours",
+          "type": "INT",
+          "search": true,
+          "storeDocValues": true
+        },
+        {
+          "name": "zipcode",
+          "type": "ATOM",
+          "search": true,
+          "storeDocValues": true
+        }
+      ]
+    }
```
Might be good to try a simple nested object field too.

taoyyu commented:
Addressed the comments and added tests for flattened nested docs. Changed the interface to support child documents in the future.

aprudhomme commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/NumberFieldDef.java:
```diff
@@ -181,14 +181,18 @@ protected FacetValueType parseFacetValueType(Field requestField) {
    */
   protected abstract Number getSortMissingValue(boolean missingLast);
 
+  protected Number parseNumberString(String numberString) {
```
nit: doc string

aprudhomme reviewed (approved):
You will need to merge master and rebuild the grpc gateway files.

taoyyu commented:
added some code to support indexing nested object with child documents.
addressed some comments and pulled master.
Query code will be in a separate PR. Will not add new logic/features in this pr

aprudhomme commented on clientlib/src/main/proto/yelp/nrtsearch/luceneserver.proto:
```diff
@@ -432,7 +433,9 @@ message Field {
     repeated Field childFields = 26; // Child fields accessible by dot notation, index same data as parent
+
     bool eagerGlobalOrdinals = 27; // Compute facet global ordinals for this field up front, otherwise this is done lazily on first query. Currently only for SORTED_SET_DOC_VALUES facet type.
+    bool nested = 28; // True if object data should be indexed as child documents
```
Nested is a bit overloaded. Maybe nestedDoc? Or a NESTED FieldType.

taoyyu commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/AddDocumentHandler.java:
```diff
@@ -124,19 +161,41 @@ public long runIndexingJob() throws Exception {
+          DocumentsContext documentsContext =
+              AddDocumentHandler.LuceneDocumentBuilder.getDocumentsContext(
+                  addDocumentRequest, indexState);
+          if (documentsContext.hasNested()) {
+            try {
+              if (idFieldDef != null) {
+                updateNestedDocuments(documentsContext, idFieldDef, shardState);
+              } else {
+                addNestedDcouments(documentsContext, shardState);
+              }
+            } catch (IOException e) { // This exception should be caught in parent to and set
+              // responseObserver.onError(e) so client knows the job failed
+              logger.warn(
+                  String.format(
+                      "ThreadId: %s, IndexWriter.addDocuments failed",
+                      Thread.currentThread().getName() + Thread.currentThread().getId()));
+              throw new IOException(e);
+            }
+          } else {
+            documents.add(documentsContext.getRootDocument());
+          }
```
This is special requirements to use ToParentBlockJoinQuery which is supposed to handle child docs https://lucene.apache.org/core/4_4_0/join/org/apache/lucene/search/join/ToParentBlockJoinQuery.html

taoyyu commented on src/main/java/com/yelp/nrtsearch/server/luceneserver/field/ObjectFieldDef.java:
```diff
@@ -0,0 +1,130 @@
+   */
+  private Document createChildDocument(
+      Map<String, Object> fieldValue, List<List<String>> facetHierarchyPaths) {
+    Document document = new Document();
```
yea, I'll add them later when I am adding the query. It will be more clear whether these identifier are needed when there is a ToParentBlockJoinQuery test
